### PR TITLE
Recode bilibili

### PR DESCRIPTION
**Pull request by Kotoki1337**
solved #2666 

Only important pages are implemented, in order to allow people who download presences in the presences store to use presence again.
I will add other pages later (in other pull requests).

image

### CODE REVIEW

File: websites/B/bilibili/presence.ts [hunk 1]
```diff
@@ -1,501 +1,117 @@
...
-    );
-    presenceData.smallImageKey = videoPaused ? "pause" : "play";
-    presenceData.smallImageText = videoPaused
-      ? (await strings).pause
-      : (await strings).play;
     presenceData.startTimestamp = timestamps[0];
     presenceData.endTimestamp = timestamps[1];
 
+    presenceData.smallImageKey = videoPaused ? "pause" : "play";
+    // presenceData.smallImageText = videoPaused ? (await strings).pause : (await strings).play; Not working due language issues, this issue should fix in extension
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:48]: Remove comment 
```suggestion
```

File: websites/B/bilibili/presence.ts [hunk 2]
```diff
@@ -1,501 +1,117 @@
...
-        " - 搜索结果 - 哔哩哔哩弹幕视频网 - ( ゜- ゜)つロ  乾杯~  - bilibili",
-        ""
-      );
-      presenceData.smallImageKey = "search";
+  }
+
+  switch (document.location.hostname) {
+    case "www.bilibili.com": {
+      switch (urlpath[1]) {
+        case "": {
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:59]: Change this to the default on the switch

File: websites/B/bilibili/presence.ts [hunk 3]
```diff
@@ -1,501 +1,117 @@
...
   }
 
-  if (presenceData.details == null) {
+  if (presenceData.details === null) {
     presence.setTrayTitle();
     presence.setActivity();
-  } else {
-    presence.setActivity(presenceData);
-  }
+  } else presence.setActivity(presenceData);
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:116]:
```suggestion presence.setActivity(presenceData);
```

File: websites/B/bilibili/presence.ts [hunk 4]
```diff
@@ -1,501 +1,117 @@
...
+        }
+      ];
+      break;
     }
   }
 
-  if (presenceData.details == null) {
+  if (presenceData.details === null) {
     presence.setTrayTitle();
     presence.setActivity();
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:497]: Remove this and just have presence.setActivity(presenceData)

File: websites/B/bilibili/dist/metadata.json [hunk 5]
```diff
@@ -1,14 +1,8 @@
 {
   "author": {
-    "name": "Bas950",
-    "id": "241278257335500811"
```
Comment (EncryptedDev) [inline websites/B/bilibili/dist/metadata.json:4]: Move Bas to contributor

File: websites/B/bilibili/presence.ts [hunk 6]
```diff
@@ -1,501 +1,117 @@
 const presence = new Presence({ clientId: "639591760791732224" }),
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:1]: If you're now the author, you own the clientId

File: websites/B/bilibili/dist/metadata.json [hunk 7]
```diff
@@ -1,14 +1,8 @@
...
+    "name": "Moe Kotoki",
+    "id": "276627005280092162"
   },
```
Comment (EncryptedDev) [inline websites/B/bilibili/dist/metadata.json:5]: Have you checked with Bas about the author change?

File: websites/B/bilibili/presence.ts [hunk 8]
```diff
@@ -1,501 +1,117 @@
...
-        video = document.querySelector(
-          "#bilibiliPlayer > div.bilibili-player-area.video-state-blackside.progress-shadow-show > div.bilibili-player-video-wrap > div.bilibili-player-video > video"
-        );
-      }
-
-      videoDuration = video.duration;
-      videoCurrentTime = video.currentTime;
+  async function getTimestamps() {
+    video = document.querySelector("bwp-video");
+    if (video === null) {
```
Comment (EncryptedDev) [inline websites/B/bilibili/presence.ts:24]:
```suggestion
    if (!video) {
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Seems like a lot of functionality has been removed, why is this?

**Comment by Kotoki1337** [comment]
> Seems like a lot of functionality has been removed, why is this?

The previous code was too messy, and the bilibili webpage also changed a lot, needed to be recode

**Comment by EncryptedDev** [comment]
> > Seems like a lot of functionality has been removed, why is this?
> 
> The previous code was too messy, and the bilibili webpage also changed a lot, needed to be recode

I totally understand that, but it seems like you removed a tonne of functionality on top of that? Did the service deprecate most functionality or does the presence support less functionality? 

If you want to recode any presence, you cannot reduce the amount of pages/functions said service supports, you must still support everything. You just need to improve code quality without reducing functionality.

**Comment by Kotoki1337** [comment]
> > > Seems like a lot of functionality has been removed, why is this?
> 
> > 
> 
> > The previous code was too messy, and the bilibili webpage also changed a lot, needed to be recode
> 
> 
> 
> I totally understand that, but it seems like you removed a tonne of functionality on top of that? Did the service deprecate most functionality or does the presence support less functionality? 
> 
> 
> 
> If you want to recode any presence, you cannot reduce the amount of pages/functions said service supports, you must still support everything. You just need to improve code quality without reducing functionality. 

I understand what u thought, I want to merge the basic features first. Too many people are coming to me and asking me to fix the presence. If I want to support everything, I'll probably need a few weeks or even a month. So let them use the simplest available one first, and then i gonna add the rest of the features later.

**Comment by EncryptedDev** [comment]
> > > > Seems like a lot of functionality has been removed, why is this?
> > 
> > 
> > > 
> > 
> > 
> > > The previous code was too messy, and the bilibili webpage also changed a lot, needed to be recode
> > 
> > 
> > I totally understand that, but it seems like you removed a tonne of functionality on top of that? Did the service deprecate most functionality or does the presence support less functionality?
> > If you want to recode any presence, you cannot reduce the amount of pages/functions said service supports, you must still support everything. You just need to improve code quality without reducing functionality.
> 
> I understand what u thought, I want to merge the basic features first. Too many people are coming to me and asking me to fix the presence. If I want to support everything, I'll probably need a few weeks or even a month. So let them use the simplest available one first, and then i gonna add the rest of the features later.

Ah, I see, makes more sense. Let's see what a maintainer would do here though

**Comment by Slowlife01** [comment]
No draft, if you are going to add it in this pull request please do it within 7 days if not the current code is fine by me just have to address Rhys's requests

**Comment by Kotoki1337** [comment]
> No draft, if you are going to add it in this pull request please do it within 7 days if not the current code is fine by me just have to address Rhys's requests

I'm draft because I'm not at home at the moment to deal with Rhys's review, because I was afraid that someone else would continue to review and cause Rhys's review to be outdated, but I'm home now and I'll be able to finish this pull request soon

File: websites/B/bilibili/presence.ts [hunk 9]
```diff
@@ -1,501 +1,113 @@
...
-      presenceData.state = "of: " + user.innerText;
-      presenceData.smallImageKey = "reading";
-    } else if (document.location.pathname.includes("/topic")) {
-      title = document.querySelector(
-        "#app > div.page-container.p-rel > div.top-header.p-rel > div.tag-title-content.fs-28.ls-0 > div.tag-title.d-i-block"
-      );
-      presenceData.startTimestamp = browsingStamp;
-      presenceData.details = "Viewing dynamic";
-      presenceData.state = "Tag: " + title.innerText;
+      timestamps = presence.getTimestamps(video.currentTime, video.duration);
```
Comment (Bas950) [inline websites/B/bilibili/presence.ts:27]: use getmediatimestamps

**Review by Bas950 — CHANGES_REQUESTED** [review]

**Review by Slowlife01 — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**